crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1748,8 +1748,37 @@ impl<'db> TypeInferenceBuilder<'db> {
         self.infer_arguments(arguments);
         self.infer_expression(func);
 
-        // TODO resolve to return type of `func`, if its a callable type
-        Type::Unknown
+        match func.as_ref() {
+            name_expr @ ast::Expr::Name(..) => {
```

Comment:
We shouldn't be matching on the syntactic form of `func` here; the inner match on the type of `func` should really be the only one here. And we don't need to go digging into `self.types.expressions` for that type. `self.infer_expression(func)` already returns the inferred type for `func` (we just aren't using it), and then we can just match on that type. It shouldn't matter to us here whether `func` is a simple `Name`, or an attribute expression, or whatever else, those details are all handled by `infer_expression`; all that matters here is the type we infer for it.

Reply:
Ah, excellent. 💯